ivadomed/utils.py

```diff
@@ -231,33 +235,60 @@ def structurewise_uncertainty(fname_lst, fname_hard, fname_unc_vox, fname_out):
     data_iou, data_cv, data_avgUnc = np.zeros(data_hard.shape), np.zeros(data_hard.shape), np.zeros(data_hard.shape)
 
     # load all MC simulations and label them
-    data_lst, data_l_lst = [], []
+    data_lst = []
+    data_l_arr = np.array()
     for fname in fname_lst:
         nib_im = nib.load(fname)
         data_im = nib_im.get_fdata()
         data_lst.append(data_im)
-        data_im_l, _ = label(data_im, structure=bin_struct)
-        data_l_lst.append(data_im_l)
+        data_im_l = []
+        for i in range(data_im.shape[-1]):
+            data_im_tmp, _ = label(data_im[..., i], structure=bin_struct)
+            data_im_l.append(data_im_tmp)
+        np.append(data_l_arr, data_im_l)
         del nib_im
+    # channel went first due to the 'append' function
+    # Before the transpose below, the variable data_l_arr has the shape:
+    # MC_simulation X Channel X Height X Width X Depth
+    # After transpose, the shape becomes: MC_simulation X Height X Width X Depth X Channel
+    # this is wanted because the input of run_uncertainty are nifti_masks created by network predictions
+    # These nifti masks have this shape: Height X Width X Depth X Channel.
+    # We want an output of the same shape as the input.
+    if len(data_l_arr.shape) == 5:
+        data_l_arr = np.transpose(data_l_arr, (0, 2, 3, 4, 1))
+    elif len(data_l_arr.shape) == 4:
+        data_l_arr = np.transpose(data_l_arr, (0, 2, 3, 1))
+    else:
+        raise ValueError("Input data of shape {} is not valid. Images must be 4D or 5D".format(len(data_l_arr.shape)))
 
     # loop across all structures of data_hard_l
+
     for i_l in range(1, n_l + 1):
         # select the current structure, remaining voxels are set to zero
-        data_i_l = (data_hard_l == i_l).astype(np.int)
+        data_i_l = (np.array(data_hard_l) == i_l).astype(np.int)
+        # channel went first with 'append' function for data_hard_l in line 225/226 (because we looped over channels)
+        # before transpose shape is supposed to be Channel X Height X Width X depth
+        # After transpose shape is expected to be  Height X Width X Depth X Channel
+        if len(data_i_l.shape) == 4:
+            data_i_l = np.transpose(data_i_l, (1, 2, 3, 0))
+        elif len(data_i_l.shape) == 3:
+            data_i_l = np.transpose(data_i_l, (1, 2, 0))
+        else:
+            raise ValueError("Input data of shape {} is not valid. image must be 3D or 4D".format(len(data_l_arr.shape)))
 
         # select the current structure in each MC sample
         # and store it in data_mc_i_l_lst
         data_mc_i_l_lst = []
         # loop across MC samples
         for i_mc in range(len(data_lst)):
```

Comment:
The pythonic way to iterate over paired values from two containers is to use `zip()` (see https://docs.python.org/3/library/functions.html#zip)

Code would look cleaner, something like:
```
for img, label in zip(data_lst, data_l_arr):
    # in a nutshell, img = data_lst[i] and label = data_l_arr[i]
    data_i_inter = data_i_l * label
    i_mc_l = np.max(data_i_inter)

    if i_mc_l > 0:
        data_mc_i_l = np.copy(img)
        data_mc_i_l[label != i_mc_l] = 0.
    else:  # no structure in this sample
        data_mc_i_l = np.zeros(img.shape)
    data_mc_i_l_lst.append(data_mc_i_l)

It's purely a python thing, but makes the code more readable than having matrices with indexes all over the place.
```

Reply:
I get what you mean. It might be a bit out of scope for this PR.